### PR TITLE
Add WP_Cache to get_authors function

### PR DESCRIPTION
**Pull request by addaeabeng**

### CODE REVIEW

File: inc/template.php [hunk 1]
```diff
@@ -52,10 +52,16 @@ function get_authors( WP_Post $post ) : array {
 	}
 
 	/** @var WP_User[] */
-	$users = get_users( [
-		'include' => $author_ids,
-		'orderby' => 'include',
-	] );
+	$users = wp_cache_get( 'authors', 'authorship' );
```
Comment (tomjn) [inline inc/template.php:55]: @addaeabeng this means all posts will share the same cached users, meaning it won't be possible to have more than one post on a site, the cache needs to be per-post

File: tests/phpunit/test-rest-api-post-property.php [hunk 2]
```diff
@@ -18,6 +18,11 @@
 use WP_REST_Request;
 
 class TestRESTAPIPostProperty extends RESTAPITestCase {
+
+	public function setUp() {
+		wp_cache_flush();
+	}
+
```
Comment (mattheu) [inline tests/phpunit/test-rest-api-post-property.php:25]: Given you've not added/changed any tests, can this be removed.

**Review by mattheu — CHANGES_REQUESTED** [review]
You're saying that checks are failing... you do need to fix this. 

One is a coding standards issue and can easily be fixed: https://github.com/humanmade/authorship/runs/5391943454?check_suite_focus=true

The other is indeed a failing unit test. Now this indicates to me that something you have done here has broken things. Looking at the test that it is failing, it is updating the authors. I suspect that the issue is you're getting cached data. 

Probably the only way to avoid this is to clear the caches for a post when you update it (see `set_authors`).

File: tests/phpunit/test-rest-api-post-property.php [hunk 3]
```diff
@@ -18,6 +18,7 @@
 use WP_REST_Request;
 
 class TestRESTAPIPostProperty extends RESTAPITestCase {
+
```
Comment (mattheu) [inline tests/phpunit/test-rest-api-post-property.php:21]: Still need to remove this new line from this file.

File: inc/template.php [hunk 4]
```diff
@@ -47,16 +57,27 @@ function get_author_ids( WP_Post $post ) : array {
...
 function get_authors( WP_Post $post ) : array {
 	$author_ids = get_author_ids( $post );
+
 	if ( empty( $author_ids ) ) {
 		return [];
 	}
 
-	/** @var WP_User[] */
+	// Just a very simple way to convert author IDs into a string that can be used as a cache key.
+	// Could also generate a hash such as hash( 'crc32', json_encode( $author_ids ) ); .
```
Comment (mattheu) [inline inc/template.php:66]: My apologies - I think you can just remove this comment completely. I added them mostly as a suggestion to you, it's ultimately your call whether just `implode` is sufficient or whether to use a hashing function. But either way you need to clean up this

File: inc/template.php [hunk 5]
```diff
@@ -170,6 +191,8 @@ function set_authors( WP_Post $post, array $authors ) : array {
 		throw new Exception( $terms->get_error_message() );
 	}
 
+	wp_cache_flush();
```
Comment (mattheu) [inline inc/template.php:194]: I think flushing the _entire_ cache is too heavy handed. Remember this is a plugin used by many sites! As much as cache _should_ be ephemeral, I think we have to be cautious about clearing it. Take for example a large high traffic site with a _lot_ of cached data, clearing it when a new post is published could significantly impact the performance for subsequent requests. 

It's easy to get around though. Just delete the individual cache entry by key. 

```suggestion
	wp_cache_delete( 'author_ids_' . $post->ID, 'authorship' );
```

**Review by mattheu — DISMISSED** [review]
Still a couple of changes necessary here. At least all the checks are passing though!

File: tests/phpunit/test-rest-api-post-property.php (hunk 2 above)
Comment (tomjn) [reply tests/phpunit/test-rest-api-post-property.php]: @addaeabeng I see we kept this, can you add an explainer here on why the feedback was dismissed without changes?

**Review by tomjn — DISMISSED** [review]

**Comment by tomjn** [comment]
This appears good, but I'd like someone who's worked with this plugin that isn't on our project to review it too. Once that's done, a 0.3.0 release might be necessary but we can always reference a specific commit at our end

**Review by johnbillion — CHANGES_REQUESTED** [review]
I'd like to know the performance benefit this provides. A description for the PR would be great.

* `get_author_ids()` calls `wp_get_post_terms()` which ultimately calls `WP_Term_Query::get_terms()` which contains a caching layer. Multiple subsequent calls to `get_author_ids()` returns cached data and does not perform additional database queries.
* `get_authors()` does perform a user query each time, but as it's a query for a list of user IDs it's very fast. Caching complete `WP_User` objects risks stale data without comprehensive cache flushing when users are updated.

**Review by shadyvb — CHANGES_REQUESTED** [review]
Based on John's review, I don't think we should proceed with the additional caching layer if one already exists. Not all consumers would be interested in such aggressive caching. Some benchmarks before/after would be nice to support the argument.

Consumers can always proxy the functions/classes as needed to add more strict caching requirements.

**Comment by mattheu** [comment]
> get_authors() does perform a user query each time, but as it's a query for a list of user IDs it's very fast. Caching complete WP_User objects risks stale data without comprehensive cache flushing when users are updated.

Yes, this the actual reason we flagged this for investigation. Hypothetically a page may have 10 posts shown, and `get_authors` is called 10 times. If they all have the exact same authors set for each, then this leads to the same identical query being run 10 times. This was flagged because we did an audit and were seeing these duplicate queries in Query Monitor. In our case we're actually doing some additional sorting of in user queries, which makes it less performant than a simple query by IDs.

Really I had hoped to avoid too aggressive caching here, and probably fine to just be be cached for the duration of the request. But understand the concerns. It's not so easy to clear the cache. Not sure how to work around this really.

**Comment by johnbillion** [comment]
The result could go into a static variable inside `get_authors()`, maybe an array keyed by post ID. Then it would just be cached for the duration of the request.

**Comment by mattheu** [comment]
Made an adjustment to this to cache the user query. I am keen to avoid keying this by ID as this only solves the issue of calling multiple times for the same post. And not the issue of having different posts with the same authors. (Admittedly this is more an issue for my dev environment where all posts are written by "admin" :) )

**Comment by shadyvb** [comment]
Still not keen on the change, this is aggressively caching the list of users, and not even in object cache, so consumers wouldn't be able to override the behavior, which isn't expected. User objects can change between different function calls, and this does not count for it.

I'd rather look at a way that is less aggressive, or to use filters to possibly enable consumers to do the caching on their end.

**Comment by addaeabeng** [comment]
After further discussion with @johnbillion will consider adding a pre filter to allow external code to manage caching

**Comment by mattheu** [comment]
I think we can close this. No longer a project need for this changes and quite a bit of pushback on this ticket.

**Pull request closed without merging**